RecoMTD/TrackExtender/plugins/TrackExtenderWithMTD.cc

```diff
@@ -0,0 +1,484 @@
+  TransientTrackingRecHit::ConstRecHitContainer output;
+  const vector<const DetLayer*>& layers = geo->allETLLayers();
+
+  auto cmp = [](const unsigned one, const unsigned two) -> bool { return one < two; };
+
+  auto tTrack = builder->build(track);
+
+  for (auto ilay = layers.begin(); ilay!=layers.end(); ++ilay) {
+    const MTDRingForwardDoubleLayer* layer = (const MTDRingForwardDoubleLayer*) (*ilay);
+    const BoundDisk& disk = layer->specificSurface();
+    const double diskZ = disk.position().z();
+
+    // get the outermost trajectory point on the track    
+    TrajectoryStateOnSurface tsos = tTrack.outermostMeasurementState();
+    if( tsos.globalPosition().z() * diskZ < 0 ) continue; // only propagate to the disk that's on the same side
+
+    PropagatorWithMaterial prop(anyDirection,0.13957018,field,1.6,false,0.1,true);
+
+    pair<bool, TrajectoryStateOnSurface> comp = layer->compatible(tsos,prop,*theEstimator);
+    if( comp.first ) {
+      // if we're compatible with the disc, try to find modules
+
+      vector<DetLayer::DetWithState> compDets = layer->compatibleDets(tsos,prop,*theEstimator);
+      const bool hasDets = !compDets.empty();
+      if( hasDets ) {
+	for( const auto& detWithState : compDets ) {
+	  auto range = hits.equal_range(detWithState.first->geographicalId(),cmp);	  
+	  for( auto detitr = range.first; detitr != range.second; ++detitr ) {
+	    auto best = detitr->end();
+	    double best_chi2 = std::numeric_limits<double>::max();
+	    for( auto itr = detitr->begin(); itr != detitr->end(); ++itr ) {
+	      auto est =  theEstimator->estimate(detWithState.second,*itr);
+	      if( est.first && est.second < best_chi2 ) { // just take the best chi2
+		best = itr;
+		best_chi2 = est.second;
+	      }
+	    }
+	    if( best != detitr->end() ) {
+	      output.push_back(hitbuilder->build(&*best));
+	    }
+	  }
+	  
+	}
+      }
+    }
+  }
+  return output;
+}
+
+
+//below is unfortunately ripped from other places but 
+//since track producer doesn't know about MTD we have to do this
+template<class TrackCollection>
+reco::Track TrackExtenderWithMTDT<TrackCollection>::buildTrack(const reco::Track& orig,
+							       const Trajectory& traj,
+							       const reco::BeamSpot& bs,
+							       const MagneticField* field,
+							       bool hasMTD) const {  
+  // get the state closest to the beamline
+  TrajectoryStateOnSurface stateForProjectionToBeamLineOnSurface = 
+    traj.closestMeasurement(GlobalPoint(bs.x0(),bs.y0(),bs.z0())).updatedState();
+  
+  if UNLIKELY(!stateForProjectionToBeamLineOnSurface.isValid()){
+    edm::LogError("CannotPropagateToBeamLine")<<"the state on the closest measurement isnot valid. skipping track.";
+    return reco::Track();
+  }
+
+  const FreeTrajectoryState & stateForProjectionToBeamLine=*stateForProjectionToBeamLineOnSurface.freeState();
+
+  PropagatorWithMaterial thePropagator(anyDirection,0.13957018,field,1.6,false,0.1,true);
```

Comment:
magic numbers again

Reply:
I need to take the propagator from the ES anyway. Must less overhead and churn over all that way.